google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java

```diff
@@ -424,24 +427,55 @@ public String updateCallback(Transaction transaction) {
 
   @Test
   public void successfulTransaction() throws Exception {
-    final DocumentReference documentReference = addDocument("foo", 1);
-
-    String result =
-        firestore
-            .runTransaction(
-                new Transaction.Function<String>() {
-                  @Override
-                  public String updateCallback(Transaction transaction)
-                      throws ExecutionException, InterruptedException {
-                    DocumentSnapshot documentSnapshot = documentReference.get().get();
-                    documentReference.update("foo", documentSnapshot.getLong("foo") + 1);
-                    return "bar";
-                  }
-                })
-            .get();
-
-    assertEquals("bar", result);
-    assertEquals(2L, documentReference.get().get().get("foo"));
+    final DocumentReference documentReference = addDocument("counter", 1);
+
+    final Semaphore firstLock = new Semaphore(0);
+    final Semaphore secondLock = new Semaphore(0);
+
+    final AtomicInteger attempts = new AtomicInteger();
+
+    // One of these transaction fails and has to be retried since they both acquire locks on the
+    // same document, which they then modify.
+    ApiFuture<String> firstTransaction =
+        firestore.runTransaction(
+            new Transaction.Function<String>() {
+              @Override
+              public String updateCallback(Transaction transaction)
+                  throws ExecutionException, InterruptedException {
+                attempts.incrementAndGet();
+
+                DocumentSnapshot documentSnapshot = transaction.get(documentReference).get();
+                firstLock.release();
+                secondLock.acquire();
+                transaction.update(
+                    documentReference, "counter", documentSnapshot.getLong("counter") + 1);
+                secondLock.release(); // Make sure we don't lock on retry
+                return "foo";
+              }
+            });
+
+    ApiFuture<String> secondTransaction =
+        firestore.runTransaction(
+            new Function<String>() {
+              @Override
+              public String updateCallback(Transaction transaction)
+                  throws ExecutionException, InterruptedException {
+                attempts.incrementAndGet();
+
+                DocumentSnapshot documentSnapshot = transaction.get(documentReference).get();
+                firstLock.acquire();
+                secondLock.release();
```

Comment:
nit: can we release before acquire?

My reasoning is if both transactions release before acquire, the synchronization should still work, but if both acquire before release, we'll deadlock. I think release-before-acquire makes it easier to prove deadlock-freedom in our heads.

Reply:
Updated this to use the much more suitable CountDownLatch.